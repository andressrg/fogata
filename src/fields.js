const { ValidationError } = require('./errors');

class Validator {
  validate() {}
}

class RequiredValidator extends Validator {
  validate(value) {
    super.validate(...arguments);
    if (value === null || value === undefined) {
      throw new ValidationError('Required value.');
    }
  }
}

class Field {
  constructor(params = {}) {
    this.validators = params.validators || [];

    this.required = params.required;
    if (this.required === true) {
      this.validators = [...this.validators, new RequiredValidator()];
    }
  }

  load(input) {
    const errors = this.validators
      .map(validator => {
        try {
          validator.validate(input);
        } catch (err) {
          return err.toString();
        }
      })
      .filter(e => !!e);

    if (errors.length > 0) {
      throw new ValidationError(errors);
    }

    return input;
  }

  dump(input) {
    return input;
  }
}

class StringField extends Field {
  load(input) {
    input = super.load(...arguments);
    return `${input}`;
  }

  dump(input) {
    input = super.dump(...arguments);
    return this.load(input);
  }
}

class DateField extends Field {
  load(input) {
    input = super.load(...arguments);
    return new Date(input);
  }

  dump(input) {
    input = super.dump(...arguments);
    return input.toISOString();
  }
}

class EmailValidator extends Validator {
  validate(value) {
    super.validate(...arguments);

    if (value !== undefined && !value.includes('@')) {
      throw new ValidationError(`${value} is not a valid email address.`);
    }
  }
}

class EmailField extends Field {
  constructor() {
    super();
    this.validators = [...this.validators, new EmailValidator()];
  }
}

class NestedField extends Field {
  constructor(nestedSchema) {
    super(...arguments);
    this.nestedSchema = nestedSchema;
  }

  load(input) {
    input = super.load(...arguments);
    return this.nestedSchema.load(input);
  }

  dump(input) {
    input = super.dump(...arguments);
    return this.nestedSchema.dump(input);
  }
}

module.exports = {
  Field,
  StringField,
  NestedField,
  DateField,
  EmailField
};
