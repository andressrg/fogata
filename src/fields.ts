import { ValidationError } from './errors';
import { Schema } from './index';

export class Validator implements Validator {
  validate(value: any) {}
}

class RequiredValidator extends Validator {
  validate(value: any) {
    super.validate(value);
    if (value === null || value === undefined) {
      throw new ValidationError('Required value.');
    }
  }
}

export interface FieldParams {
  validators?: Array<Validator>;
  required?: boolean;
}

export class Field {
  validators: Validator[];
  required: boolean;

  constructor(params: FieldParams = {}) {
    this.validators = params.validators || [];

    this.required = params.required || false;
    if (this.required === true) {
      this.validators = [...this.validators, new RequiredValidator()];
    }
  }

  load(input: any): any {
    const errors: string[] = this.validators
      .map(validator => {
        try {
          validator.validate(input);
        } catch (err) {
          return err.toString();
        }
      })
      .filter(e => !!e);

    if (errors.length > 0) {
      throw new ValidationError(errors);
    }

    return input;
  }

  dump(input: any): any {
    return input;
  }
}

export class StringField extends Field {
  load(input: any): string {
    input = super.load(input);
    return `${input}`;
  }
}

export class DateField extends Field {
  load(input: any): Date {
    input = super.load(input);
    return new Date(input);
  }

  dump(input: Date): string {
    input = super.dump(input);
    return input.toISOString();
  }
}

export class EmailValidator extends Validator {
  validate(value: string) {
    super.validate(value);

    if (value !== undefined && !value.includes('@')) {
      throw new ValidationError(`${value} is not a valid email address.`);
    }
  }
}

export class EmailField extends Field {
  constructor(params: FieldParams = {}) {
    super(params);
    this.validators = [...this.validators, new EmailValidator()];
  }
}

export class NestedField extends Field {
  nestedSchema: Schema;

  constructor(nestedSchema: Schema) {
    super(...arguments);
    this.nestedSchema = nestedSchema;
  }

  load(input: any): any {
    input = super.load(input);
    return this.nestedSchema.load(input);
  }

  dump(input: any) {
    input = super.dump(input);
    return this.nestedSchema.dump(input);
  }
}
