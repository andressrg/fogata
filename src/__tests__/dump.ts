import * as _ from 'lodash';
import { Schema } from '../';
import * as fields from '../fields';

class MyAsyncField extends fields.Field {
  async load(input: any) {
    return await Promise.resolve(input);
  }

  async dump(input: any) {
    return await Promise.resolve(input);
  }
}

class ArtistSchema extends Schema {
  static artistName = new fields.StringField();
  static myAsyncAttr = new MyAsyncField();
}

class AlbumSchema extends Schema {
  static title = new fields.StringField();
  static release_date = new fields.DateField();
  static artist = new fields.NestedField(new ArtistSchema());
}

class Instancifier {
  constructor(data: any) {
    Object.getOwnPropertyNames(data).forEach(attr => {
      // @ts-ignore
      this[attr] = data[attr];
    });
  }
}

it('loads', async () => {
  expect(
    await new AlbumSchema().load({
      artist: { artistName: 'El Cuarteto De Nos', myAsyncAttr: 'something' },
      title: 'Habla Tu Espejo',
      release_date: '2014-10-06T18:51:17.749Z'
    })
  ).toMatchSnapshot();
});

it('dumps', async () => {
  expect(
    await new AlbumSchema().dump(
      new Instancifier({
        artist: new Instancifier({
          artistName: 'El Cuarteto De Nos',
          myAsyncAttr: 'something'
        }),
        title: 'Habla Tu Espejo',
        release_date: new Date('2014-10-06T18:51:17.749Z')
      })
    )
  ).toMatchSnapshot();
});

it('dumps only', async () => {
  expect(
    _.keys(
      await new AlbumSchema({ only: ['title', 'release_date'] }).dump(
        new Instancifier({
          artist: new Instancifier({
            artistName: 'El Cuarteto De Nos',
            myAsyncAttr: 'something'
          }),
          title: 'Habla Tu Espejo',
          release_date: new Date('2014-10-06T18:51:17.749Z')
        })
      )
    )
  ).toEqual(['title', 'release_date']);
});

it('dumps exclude', async () => {
  expect(
    _.keys(
      await new AlbumSchema({ exclude: ['title'] }).dump(
        new Instancifier({
          artist: new Instancifier({
            artistName: 'El Cuarteto De Nos',
            myAsyncAttr: 'something'
          }),
          title: 'Habla Tu Espejo',
          release_date: new Date('2014-10-06T18:51:17.749Z')
        })
      )
    )
  ).toEqual(['release_date', 'artist']);
});

it('dumps many', async () => {
  const result = await new AlbumSchema({ many: true }).dump(
    _.times(
      3,
      () =>
        new Instancifier({
          artist: new Instancifier({
            artistName: 'El Cuarteto De Nos',
            myAsyncAttr: 'something'
          }),
          title: 'Habla Tu Espejo',
          release_date: new Date('2014-10-06T18:51:17.749Z')
        })
    )
  );
  expect(result).toBeInstanceOf(Array);
  expect(result).toHaveLength(3);
  expect(result).toMatchSnapshot();
});
