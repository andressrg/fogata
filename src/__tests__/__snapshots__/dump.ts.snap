// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dumps 1`] = `
Object {
  "artist": Object {
    "artistName": "El Cuarteto De Nos",
    "myAsyncAttr": "something",
  },
  "release_date": "2014-10-06T18:51:17.749Z",
  "title": "Habla Tu Espejo",
}
`;

exports[`dumps many 1`] = `
Array [
  Object {
    "artist": Object {
      "artistName": "El Cuarteto De Nos",
      "myAsyncAttr": "something",
    },
    "release_date": "2014-10-06T18:51:17.749Z",
    "title": "Habla Tu Espejo",
  },
  Object {
    "artist": Object {
      "artistName": "El Cuarteto De Nos",
      "myAsyncAttr": "something",
    },
    "release_date": "2014-10-06T18:51:17.749Z",
    "title": "Habla Tu Espejo",
  },
  Object {
    "artist": Object {
      "artistName": "El Cuarteto De Nos",
      "myAsyncAttr": "something",
    },
    "release_date": "2014-10-06T18:51:17.749Z",
    "title": "Habla Tu Espejo",
  },
]
`;

exports[`loads 1`] = `
Object {
  "artist": Object {
    "artistName": "El Cuarteto De Nos",
    "myAsyncAttr": "something",
  },
  "release_date": 2014-10-06T18:51:17.749Z,
  "title": "Habla Tu Espejo",
}
`;
